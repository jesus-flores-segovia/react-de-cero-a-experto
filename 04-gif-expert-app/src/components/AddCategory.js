import { useState } from "react"
import PropTypes from "prop-types";

export const AddCategory = ({setCategories}) => {
    
    // Es importante inicializar el state, para evitar el warning que aparece por consola, que nos avisa de cambios incontrolados en el input
    const [inputValue, setInputValue] = useState();
    
    const handleInputChange = (e) => {
        setInputValue(e.target.value);
    };

    const handleSubmit = (e) => {
        e.preventDefault();
        if(inputValue.trim().length !== 0){
            // Recibimos las categories en el callback de la funciÃ³n del useState
            setCategories(categories => [...categories, inputValue]);
            setInputValue("");
        }
        
    };

    return (
        <form onSubmit={handleSubmit}>
            <input
                type="text"
                value={inputValue}
                onChange={handleInputChange}
            />
        </form>
    )
}

AddCategory.propTypes = {
    setCategories: PropTypes.func.isRequired
};