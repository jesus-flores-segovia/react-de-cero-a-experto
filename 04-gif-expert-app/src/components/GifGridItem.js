import PropTypes from "prop-types";

export const GifGridItem = ({id, title, url}) => {

    return (
        <div className="card animate__animated animate__fadeInLeft">
            <img src={url} alt={title}/>
            <p>{title}</p>
        </div>
    )
};

GifGridItem.propTypes = {
    id: PropTypes.string,
    title: PropTypes.string.isRequired,
    url: PropTypes.string.isRequired
};