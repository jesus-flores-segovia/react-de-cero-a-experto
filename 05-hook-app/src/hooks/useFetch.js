import { useState, useEffect, useRef } from "react";

export const useFetch = (url) => {
    
    const isMounted = useRef(true);
    const [state, setState] = useState({data: null, loading: true, error: null});

    useEffect(() => {
        return () => {
            isMounted.current = false;
            console.log("isMounted to false.");
        }
    }, [])

    useEffect(() => {
        console.log("useEffect of [url].");
        setState({data: null, loading: true, error: null});

        fetch(url)
            .then(res => res.json())
            .then(data => {
                
                    setTimeout(() => {
                        if(isMounted.current){
                            setState({
                                loading: false,
                                error: null,
                                data
                            });
                        } else{
                            console.log("setState not called.");
                        }
                    }, 4000);
            });

    }, [url]);

    return state;
}
