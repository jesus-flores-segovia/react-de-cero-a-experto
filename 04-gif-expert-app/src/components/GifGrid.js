import { useState, useEffect } from "react";
import { GifGridItem } from "./GifGridItem";

export const GifGrid = ({category}) => {
    
    const [images, setImages] = useState([]);

    useEffect(() => {
        getGifs();
    }, []);

    const getGifs = async() => {

        const url = "https://api.giphy.com/v1/gifs/search?api_key=H1fsG7dJKwEzuWy3mD6lXASZNOIU0c0y&q=Rick and Morty&limit=10";
        const response = await fetch(url);
        const {data} = await response.json();

        const gifs = data.map(img => {
            return {
                id: img.id,
                title: img.title,
                url: img.images?.downsized_medium.url
            }
        })

        setImages(gifs);
    };

    return (
        <div>
            <h3>{category}</h3>
            {
                images.map(img => (
                    <GifGridItem
                        key={img.id}
                        {...img}
                    />
                ))
            }
        </div>
    )
}
