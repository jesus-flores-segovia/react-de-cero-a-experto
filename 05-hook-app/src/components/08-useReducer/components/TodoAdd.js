import React from 'react';
import {useForm} from '../../../hooks/useForm';

export const TodoAdd = ({handleAdd}) => {

    const [{description}, handleInputChanges, reset] = useForm({
        description: ""
    });

    const handleSubmit = (e) => {
        e.preventDefault();

        if(description.trim().length === 0){
            reset();
            return;
        }

        handleAdd(description);
        reset();
    };

    return (
        <>
        <h4>Add TODO</h4>
        <hr/>
                    
        <form onSubmit={handleSubmit}>
            <input type="text" name="description" className="form-control" placeholder="Learn..." autoComplete="off" onChange={handleInputChanges} value={description}/>
            <div className="d-grid">
                <button type="submit" className="btn btn-outline-primary mt-1">Add</button>
            </div>
        </form>
        </>
    )
}
