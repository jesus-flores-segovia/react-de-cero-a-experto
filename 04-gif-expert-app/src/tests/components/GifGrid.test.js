import { shallow } from "enzyme";
import { GifGrid } from "../../components/GifGrid";
import "@testing-library/jest-dom";
import { useFetchGifs } from "../../hooks/useFetchGifs";
jest.mock("../../hooks/useFetchGifs");

describe("Tests inside file GifGrid.test.js", () => {

    const category = "Dragon Ball";

    beforeEach(() => {
        jest.clearAllMocks();
    });

    test("'GifGrid' component must show correctly", () => {
        useFetchGifs.mockReturnValue({
            data: [],
            loading: true
        });

        const wrapper = shallow(<GifGrid category={category}></GifGrid>);
        expect(wrapper).toMatchSnapshot();
    });

    test("'GifGrid' component must show items obtained from useFetchGifs custom hook", () => {
        const gifs = [
            {
                id: "aAbax5anloMNk6TSP9",
                url: "https://giphy.com/gifs/TOEIAnimationUK-goku-super-saiyan-ultra-instinct-aAbax5anloMNk6TSP9",
                title: "Dragon Ball GIF by TOEI Animation UK"
            }
        ];

        useFetchGifs.mockReturnValue({
            data: gifs,
            loading: false
        });

        const wrapper = shallow(<GifGrid category={category}></GifGrid>);
        expect(wrapper.find("p").exists()).toBe(false);
        expect(wrapper.find("GifGridItem").length).toBe(gifs.length);
    });
});