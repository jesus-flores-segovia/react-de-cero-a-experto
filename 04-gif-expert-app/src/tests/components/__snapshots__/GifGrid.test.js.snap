// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests inside file GifGrid.test.js 'GifGrid' component must show correctly 1`] = `
<Fragment>
  <h3>
    Dragon Ball
  </h3>
  <p>
    Loading...
  </p>
  <div
    className="card-grid"
  />
</Fragment>
`;
